bpf/mb_xdp.c

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright © 2022 Merbridge Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "headers/helpers.h"
+#include "headers/maps.h"
+#include "headers/mesh.h"
+#include <linux/bpf.h>
+#include <linux/if_ether.h>
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/tcp.h>
+
+static __always_inline __u32 csum_add(__u32 addend, __u32 csum)
+{
+    __u32 res = csum;
+    res += addend;
+    return (res + (res < addend));
+}
+
+static __always_inline __u32 csum_sub(__u32 addend, __u32 csum)
+{
+    return csum_add(csum, ~addend);
+}
+
+static __always_inline __u16 csum_fold_helper(__u32 csum)
+{
+    __u32 r = csum << 16 | csum >> 16;
+    csum = ~csum;
+    csum -= r;
+    return (__u16)(csum >> 16);
+}
+
+static __always_inline __u16 csum_diff4(__u32 from, __u32 to, __u16 csum)
+{
+    __u32 tmp = csum_sub(from, ~((__u32)csum));
+    return csum_fold_helper(csum_add(to, tmp));
+}
+
+__section("xdp") int mb_xdp(struct xdp_md *ctx)
+{
+    void *data = (void *)(long)ctx->data;
+    void *data_end = (void *)(long)ctx->data_end;
+    struct ethhdr *eth = (struct ethhdr *)data;
+    if ((void *)(eth + 1) > data_end) {
+        return XDP_DROP;
+    }
+    if (bpf_htons(eth->h_proto) != ETH_P_IP) {
+        return XDP_PASS;
+    }
+
+    struct iphdr *iph = (struct iphdr *)(eth + 1);
+    if ((void *)(iph + 1) > data_end) {
+        return XDP_DROP;
+    }
+
+    if (iph->protocol == IPPROTO_IPIP) {
+        iph = ((void *)iph + iph->ihl * 4);
+        if ((void *)(iph + 1) > data_end) {
+            return XDP_PASS;
+        }
+    }
+
+    if (iph->protocol != IPPROTO_TCP) {
+        return XDP_PASS;
+    }
+
+    struct tcphdr *tcph = (struct tcphdr *)(iph + 1);
+    if ((void *)(tcph + 1) > data_end) {
+        return XDP_DROP;
+    }
+    if (tcph->syn && !tcph->ack) {
+        // first packet
+        if (bpf_htons(tcph->dest) == IN_REDIRECT_PORT) {
```

Comment:
Just a nit, would it be better to use `tcph->dest == bpf_htons(IN_REDIRECT_PORT)`? I'm thinking if the compiler would do some optimization at compile time